sale_order_secondary_unit/models/sale_order.py

```diff
@@ -11,52 +12,44 @@ class SaleOrderLine(models.Model):
         "uom_field": "product_uom",
     }
 
-    secondary_uom_unit_price = fields.Float(
+    secondary_uom_price_unit = fields.Float(
         string="2nd unit price",
         digits="Product Price",
-        compute="_compute_secondary_uom_unit_price",
+        compute="_compute_secondary_uom_price_unit",
+        store=True,
+        precompute=True,
     )
 
-    product_uom_qty = fields.Float(
-        store=True, readonly=False, compute="_compute_product_uom_qty", copy=True
-    )
-
-    @api.depends("secondary_uom_qty", "secondary_uom_id", "product_uom_qty")
+    @api.depends("secondary_uom_qty", "secondary_uom_id")
     def _compute_product_uom_qty(self):
+        res = super()._compute_product_uom_qty()
         self._compute_helper_target_field_qty()
+        return res
 
     @api.onchange("product_uom")
-    def onchange_product_uom_for_secondary(self):
+    def _onchange_product_uom(self):
         self._onchange_helper_product_uom_for_secondary()
 
     @api.onchange("product_id")
-    def product_id_change(self):
-        """
-        If default sales secondary unit set on product, put on secondary
-        quantity 1 for being the default quantity. We override this method,
-        that is the one that sets by default 1 on the other quantity with that
-        purpose.
-        """
-        res = super().product_id_change()
-        line_uom_qty = self.product_uom_qty
-        self.secondary_uom_id = self.product_id.sale_secondary_uom_id
-        if self.product_id.sale_secondary_uom_id:
-            if line_uom_qty == 1.0:
-                self.secondary_uom_qty = 1.0
-                self.onchange_product_uom_for_secondary()
-            else:
-                self.product_uom_qty = line_uom_qty
-        return res
-
-    @api.depends("secondary_uom_qty", "product_uom_qty", "price_unit")
-    def _compute_secondary_uom_unit_price(self):
+    def _onchange_product_id(self):
```

Comment:
The same about the onchange.

Reply:
This one is needed to ensure smooth UX

Reply:
Could you show the UX problem you have with compute method ?

Reply:
To ensure that we're keeping the quantity in primary UoM unchanged if there's a secondary UoM on a new product and recompute the quantity in secondary UoM

Reply:
@rousseldenis do you have any does that answer your question?

Reply:
> If everything is now computed writable fields, you should convert this as well. If not, you don't know which one is going to act first (the compute or the onchange).
@pedrobaeza I'm not sure that's the case. The onchange and compute are not replacement and are called at different time. Compute is called on save and onchange can be called on an unsaved record with NewId - at least to my maybe outdated knowledge.

Reply:
That's not really true. The computation of the computed fields is delayed until they are needed, before saving or not, being finally computed on write if they haven't been yet computed (but only if they are not stored). @api.onchange methods are only executed when doing things via UI, and they are executed immediately, possible being less performant in general. That's why is convenient to move everything to computed writable when possible.

Reply:
> That's why is convenient to move everything to computed writable when possible.

Which is exactly the opposite of what we want here - the conversion between primary and secondary UoM must happen on UI changes, not on save.

Reply:
That's not what happen. If you have any field in screen that is recomputed in the computed writable methods, it's executed immediately. Please don't walk in circles and convert it to computed writable. Then, you tell me the problems you find.

Reply:
I don't walk in circles. There's a working solution implemented according to the https://github.com/OCA/product-attribute/blob/16.0/product_secondary_unit/models/product_secondary_unit_mixin.py#L9-L28 and at the moment you're requesting me to explore an approach that goes against those guidelines. Once those guidelines are changed I'd be happy to factor those changes in. Until that time I see no reason allocating time towards that, sorry.